internal/cmd/login.go

```diff
@@ -373,69 +400,113 @@ func runSignupForLogin(cli *CLI, client *api.Client) (*api.LoginRequestPasswordC
 			}
+		}
 
-			if err = promptSkipTLSVerify(cli); err != nil {
-				return nil, err
-			}
-			options.SkipTLSVerify = true
+		if err = promptVerifyTLSCert(cli, uaErr.Cert); err != nil {
+			return c, err
 		}
-	}
 
-	client := apiClient(options.Server, "", options.SkipTLSVerify)
-	return client, nil
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			return c, err
+		}
+		pool.AddCert(uaErr.Cert)
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				// set min version to the same as default to make gosec linter happy
+				MinVersion: tls.VersionTLS12,
+				RootCAs:    pool,
+			},
+		}
+		c.APIClient = apiClient(options.Server, "", transport)
+		c.TrustedCertificate = string(certs.PEMEncodeCertificate(uaErr.Cert.Raw))
+	}
+	return c, nil
 }
 
-func verifyTLS(host string) error {
-	url, err := urlx.Parse(host)
+func attemptTLSRequest(options loginCmdOptions) error {
+	reqURL := "https://" + options.Server
+	// First attempt with the system cert pool
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, reqURL, nil)
 	if err != nil {
-		logging.S.Debug("Cannot parse host", host, err)
-		logging.S.Error("Could not login. Please check the server hostname")
-		return err
+		return fmt.Errorf("failed to create request: %w", err)
 	}
-	url.Scheme = "https"
-	urlString := url.String()
 
-	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, urlString, nil)
-	if err != nil {
-		logging.S.Debugf("Cannot create request: %v", err)
-		return err
+	logging.S.Debugf("call server: test tls for %q", reqURL)
+	httpClient := http.Client{Timeout: 60 * time.Second}
+	res, err := httpClient.Do(req)
+	if err == nil {
+		res.Body.Close()
+		return nil
 	}
 
-	logging.S.Debugf("call server: test tls for %q", host)
-	res, err := http.DefaultClient.Do(req)
+	// Second attempt with an empty cert pool. This is necessary because at least
+	// on darwin, the error is the wrong type when using the system cert pool.
+	// Hopefully this will be fixed in a future version of Go.
+	req, err = http.NewRequestWithContext(context.TODO(), http.MethodGet, reqURL, nil)
 	if err != nil {
-		if !errors.As(err, &x509.UnknownAuthorityError{}) && !errors.As(err, &x509.HostnameError{}) && !strings.Contains(err.Error(), "certificate is not trusted") {
-			logging.S.Debugf("Cannot validate TLS due to an unexpected error: %v", err)
-			return err
-		}
-
-		logging.S.Debug(err)
+		return fmt.Errorf("failed to create request: %w", err)
+	}
 
-		return ErrTLSNotVerified
+	pool := x509.NewCertPool()
+	if options.TrustedCertificate != "" {
+		pool.AppendCertsFromPEM([]byte(options.TrustedCertificate))
 	}
 
-	defer res.Body.Close()
-	return nil
+	httpClient = http.Client{
+		Timeout: 60 * time.Second,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{RootCAs: pool, MinVersion: tls.VersionTLS12},
+		},
+	}
+	res, err = httpClient.Do(req)
+	urlErr := &url.Error{}
+	switch {
+	case err == nil:
+		res.Body.Close()
```

Comment:
Should we be closing the response body in any case?

Reply:
The godoc for `Client.Do` (https://pkg.go.dev/net/http#Client.Do) says this:

> On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed.

So we can only close the response body when err is nil.